sys/stdio_udp/stdio_udp.c

```diff
@@ -0,0 +1,99 @@
+/*
+ * Copyright (C) 2023 ML!PA Consulting GmbH
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     sys
+ * @{
+ *
+ * @file
+ * @brief       STDIO over UDP implementation
+ *
+ * This file implements STDIO via a UDP that can be used with e.g. netcat:
+ *
+ *      nc -u fe80::7837:fcff:fe7d:1aaf%tapbr0 2323
+ *
+ * @warning     This is entirely unsecured, only use this for debugging in
+ *              an isolated network!
+ *
+ * @author      Benjamin Valentin <benjamin.valentin@ml-pa.com>
+ *
+ * @}
+ */
+
+#include <errno.h>
+
+#include "net/sock/udp.h"
+#if IS_USED(MODULE_VFS)
+#include "vfs.h"
+#endif
+
+#ifndef CONFIG_STDIO_UDP_PORT
+#define CONFIG_STDIO_UDP_PORT       2323
+#endif
+
+#ifndef CONFIG_STDIO_UDP_RX_BUF_LEN
+#define CONFIG_STDIO_UDP_RX_BUF_LEN 64
+#endif
+
+#define MIN(a, b) ((a) > (b) ? (b) : (a))
+
+static sock_udp_t sock;
+static sock_udp_ep_t remote;
+
+void stdio_init(void)
+{
+    const sock_udp_ep_t local = {
+        .family = AF_INET6,
+        .netif = SOCK_ADDR_ANY_NETIF,
+        .port = CONFIG_STDIO_UDP_PORT,
+    };
+
+    sock_udp_create(&sock, &local, NULL, 0);
+
+#if IS_USED(MODULE_VFS)
+    vfs_bind_stdio();
+#endif
+}
+
+ssize_t stdio_read(void* buffer, size_t count)
+{
+    static uint8_t rx_buf[CONFIG_STDIO_UDP_RX_BUF_LEN];
+    static uint8_t *pos;
+    static size_t left;
+
+    /* shell will only read one byte at a time, so we buffer the input */
+    if (left == 0) {
+        int res = sock_udp_recv(&sock, rx_buf, sizeof(rx_buf),
+                                SOCK_NO_TIMEOUT, &remote);
+        if (res > 0) {
+            left = res;
+            pos = rx_buf;
+        } else {
+            return res;
+        }
+    }
+
+    count = MIN(left, count);
+    uint8_t *dst = buffer;
+    for (unsigned i = 0; i < count; ++i) {
+        *dst++ = *pos++;
+    }
```

Comment:
```suggestion
    memcpy(buffer, pos, count);
    pos += count;
```

The compiler uses builtins for `memcpy` and friends anyway, so the machine code generated will be the same. But IMO this is easier to read and it is more fun to step through when debugging.